wiki/Writing-tests-with-good-behavioural-coverage.md

```diff
@@ -0,0 +1,1265 @@
+    assertTrue(shouldBuyItem(50.0, 100.0))
+}
+
+@Test
+fun testShouldBuyItem_withPriceAboveBudget_returnsFalse() {
+    assertFalse(shouldBuyItem(150.0, 100.0))
+}
+```
+
+## 2. Writing Clear and Descriptive Test Cases
+
+Each test case should:
+
+- Clearly describe the scenario and expected outcome.
+- Use descriptive names for your test methods.
+
+Naming Convention:
+```
+testAction_withOneCondition_withSecondCondition_hasExpectedOutcome
+```
+
+Example:
+```
+testCheckSign_forPositiveInput_returnsPositive()
+testCheckSign_forNegativeInput_returnsNegative()
+testCheckSign_forZeroInput_returnsNeitherPositiveNorNegative()
+```
+
+## 3. Focusing on Specific Test Cases
+
+When testing a function that performs multiple actions, it's crucial to write test cases that focus on individual aspects of the function. This approach allows you to isolate and identify issues more effectively if a test fails, rather than dealing with a more complex failure analysis.
+
+### Why Focus on Specific Test Cases?
+Testing one thing at a time helps:
+
+- Identify Issues More Easily: If a test fails, you know precisely which aspect of the function is problematic.
+- Improve Test Clarity: Each test case has a clear purpose, making tests easier to read and maintain.
+- Isolate Failures: Helps in pinpointing issues related to a specific behavior or output.
+
+Consider a function that manages a food order process. This function does the following:
+
+1. Lists the food items.
+2. Calculates the total price.
+3. Display the order.
+4. Checks if the payment has been made and provides the corresponding message.
+
+```
+fun processOrder(order: List<String>, paymentMade: Boolean): String {
+    // List the food items
+    val itemList = order.joinToString(", ")
+    
+    // Calculate total price (mocked here for simplicity)
+    val totalPrice = order.size * 10.0
+    
+    // Display order
+    println("Order: $itemList")
+    println("Total: $totalPrice")
+    
+    // Payment status
+    return if (paymentMade) "Payment successful" else "Payment pending"
+}
+```
+
+**Potential output with payment made:**
+
+```
+Order: Pizza, Burger
+Total: 20.0
+Result: Payment successful
+```
+
+### Testing the Entire Functionality
+
+**Single Test Case:**
+
+```
+@Test
+fun testProcessOrder_allSteps_returnsCorrectMessage() {
+    val result = processOrder(listOf("Pizza", "Burger"), paymentMade = true)
+    assertThat(result).isEqualTo("Payment successful")
+}
+```
+
+**Difficulties in Testing All Aspects Together:**
+
+- Complex Failure Diagnosis: If this test fails, you need to diagnose whether the issue lies in listing items, calculating the total, displaying the order, or payment status.
+- Less Focused: It does not target individual aspects of the function, making it harder to identify which specific action failed.
+
+### Testing Specific Aspects
+
+**Test Case 1: Testing Listing items**
+
+```
+@Test
+fun testProcessOrder_listsItems_correctly() {
```

Comment:
Here & below: these test names aren't actually following the suggested naming format mentioned above very clearly. :) Can they be made more clear? Happy to think about it and offer suggestions on clearer names, if needed.

Ditto for all example test names being introduced in this PR (more than just these 4 need some work, I think).

Reply:
Yeah, should've rushed these.
 
Updated them with temporary names. Would love to hear your suggestions for more precise naming and update them later. - https://github.com/oppia/oppia-android/blob/1c1531d9d938dcee6c47078a1349d2ef2762de27/wiki/Writing-tests-with-good-behavioural-coverage.md#testing-specific-aspects

Will also check the rest of the sections to see if anything else deviates from convention.